File: src/services/ingestion/extractors/conversation-extractor.ts
import type { Document } from "../../../types/models.js";
import type { Extractor } from "./base.js";

/** Extractor for chat/conversation input. */
export class ConversationExtractor implements Extractor {
  /**
   * Normalizes conversation content to structured speaker turns.
   * @param doc Source document.
   */
  public async extract(doc: Document): Promise<string> {
    const raw = doc.rawContent.trim();

    try {
      const parsed = JSON.parse(raw) as unknown;
      if (Array.isArray(parsed)) {
        const lines = parsed
          .map((entry) => {
            if (typeof entry === "object" && entry !== null) {
              const speaker = String((entry as Record<string, unknown>).speaker ?? "unknown");
              const message = String((entry as Record<string, unknown>).message ?? "");
              return `${speaker}: ${message}`;
            }
            return null;
          })
          .filter((line): line is string => Boolean(line));

        if (lines.length > 0) {
          return lines.join("\n");
        }
      }
    } catch {
      // Fall through to line-based parsing.
    }

    const normalizedLines = raw
      .split(/\r?\n/)
      .map((line) => line.trim())
      .filter((line) => line.length > 0)
      .map((line) => {
        const match = line.match(/^([^:]{1,60}):\s*(.+)$/);
        if (!match) {
          return `unknown: ${line}`;
        }

        const speaker = match[1]?.trim() ?? "unknown";
        const message = match[2]?.trim() ?? "";
        return `${speaker}: ${message}`;
      });

    return normalizedLines.join("\n");
  }
}

File: src/services/ingestion/extractors/conversation_extractor.py
import json

from src.types.models import Document

from .base import Extractor


class ConversationExtractor(Extractor):
    """Extractor for chat/conversation input."""

    async def extract(self, doc: Document) -> str:
        """Normalizes conversation content to structured speaker turns.
        
        Args:
            doc: Source document.
            
        Returns:
            Normalized conversation text.
        """
        raw = doc.raw_content.strip()
        
        try:
            parsed = json.loads(raw)
            if isinstance(parsed, list):
                lines = []
                for entry in parsed:
                    if isinstance(entry, dict):
                        speaker = str(entry.get("speaker", "unknown"))
                        message = str(entry.get("message", ""))
                        if message:
                            lines.append(f"{speaker}: {message}")
                
                if lines:
                    return "\n".join(lines)
        except json.JSONDecodeError:
            pass  # Fall through to line-based parsing

        # Line-based parsing
        normalized_lines = []
        for line in raw.splitlines():
            line = line.strip()
            if not line:
                continue
            
            # Try to match speaker: message format
            import re
            match = re.match(r'^([^:]{1,60}):\s*(.+)$', line)
            if not match:
                normalized_lines.append(f"unknown: {line}")
            else:
                speaker = match.group(1).strip() or "unknown"
                message = match.group(2).strip()
                normalized_lines.append(f"{speaker}: {message}")
        
        return "\n".join(normalized_lines)
